README.md

[![Actions Status](https://github.com/ocamlpro/ez_text_renderer/workflows/Main%20Workflow/badge.svg)](https://github.com/ocamlpro/ez_text_renderer/actions)
[![Release](https://img.shields.io/github/release/ocamlpro/ez_text_renderer.svg)](https://github.com/ocamlpro/ez_text_renderer/releases)

# ez_text_renderer


EzTextRenderer is a library that allows to render Unicode text
to OCaml arrays representing RGB images.


* Website: https://ocamlpro.github.io/ez_text_renderer
* General Documentation: https://ocamlpro.github.io/ez_text_renderer/sphinx
* API Documentation: https://ocamlpro.github.io/ez_text_renderer/doc
* Sources: https://github.com/ocamlpro/ez_text_renderer


CHANGES.md

## v0.1.0 ( 2021-05-27 )

* Initial commit

## v0.1.1 ( 2021-06-01 )

* Switch drom skeleton from mini_lib to library


sphinx/searchindex.js
Search.setIndex({docnames:["about","index","install","license"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.cpp":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,sphinx:55},filenames:["about.rst","index.rst","install.rst","license.rst"],objects:{},objnames:{},objtypes:{},terms:{"case":3,"final":[2,3],"function":3,"long":3,"new":3,"public":3,"static":3,"switch":2,"true":3,"void":3,AND:3,And:3,BEING:3,BUT:3,But:3,FOR:3,For:3,HAS:3,NOT:3,SUCH:3,Such:3,THE:3,THERE:3,That:3,The:3,Then:2,These:3,USE:3,Use:3,WILL:3,WITH:3,abl:3,about:[1,3],abov:3,absenc:3,accept:3,access:3,accessor:3,accompani:3,achiev:3,act:3,action:3,activ:3,add:3,addit:3,address:3,advantag:3,advis:3,affect:3,after:3,aggreg:3,agre:3,agreement:3,all:[2,3],alleg:3,allow:[0,3],along:3,alreadi:3,also:3,alter:3,altern:3,although:3,among:3,ani:3,anoth:3,anyth:3,api:1,appear:3,appli:3,applic:3,appropri:3,april:3,argument:3,aris:3,arrai:0,ask:3,associ:3,assum:3,attach:3,attempt:3,attent:3,author:[1,3],automat:3,avail:2,awai:3,balanc:3,base:3,becaus:3,becom:3,been:3,being:3,believ:3,below:3,better:3,between:3,binari:3,bodi:3,boston:3,both:3,bring:3,build:1,call:[2,3],can:[2,3],cannot:3,carefulli:3,carri:3,caus:3,certain:3,chang:3,charg:3,checkout:2,choic:3,choos:3,circumst:3,claim:3,claus:3,clear:3,close:3,code:[2,3],collect:3,com:0,combin:3,compani:3,compat:3,compel:3,compet:3,compil:3,complet:3,complianc:3,compon:3,comput:3,concern:3,condit:3,consequ:3,consequenti:3,consid:3,consist:3,conspicu:3,constant:3,constitut:3,contact:3,contain:[2,3],content:3,contest:3,contradict:3,contrast:3,contribut:3,control:3,convei:3,conveni:3,coon:3,copi:3,copyright:1,correct:3,correspond:3,cost:3,could:3,count:3,countri:3,court:3,cover:3,creat:[2,3],criteria:3,custom:3,customarili:3,damag:3,data:3,date:3,david:0,debug:3,decid:3,decis:3,declerck:0,defect:3,defin:3,definit:3,deni:3,dep:2,depend:[2,3],deriv:3,design:3,detail:3,dev:2,devel:1,develop:[2,3],differ:3,direct:3,directli:3,directori:2,disadvantag:3,disclaim:3,displai:3,distinguish:3,distribut:3,distributor:3,doc:[1,2],document:[1,2,3],doe:3,donor:3,dune:1,dure:3,dynam:3,each:3,effect:3,effort:3,either:3,electron:3,els:3,employ:3,enabl:3,encourag:3,end:3,enforc:3,engin:3,ensur:3,entir:3,env:2,equival:3,especi:3,eval:2,even:3,event:3,ever:3,everi:3,everyon:3,exampl:[2,3],except:3,exchang:3,exclud:3,exclus:3,excus:3,execut:3,exercis:3,exist:3,explain:3,explan:3,explicit:3,express:3,expressli:3,extend:3,extent:3,ez_text_render:2,eztextrender:0,facil:3,fact:3,facto:3,failur:3,faith:3,fall:3,februari:3,fee:3,fifth:3,file:[2,3],find:3,first:3,fit:3,floor:3,fmt:2,follow:[2,3],forbid:3,form:3,format:2,former:3,found:3,foundat:3,franklin:3,free:3,freedom:3,frequent:3,frob:3,from:[2,3],full:3,further:3,gain:3,gave:3,gener:[2,3],geograph:3,get:3,github:1,give:3,given:3,gnu:3,goal:3,good:3,gpl:3,grant:3,grati:3,greater:3,greatest:3,guarante:3,guid:3,hacker:3,happen:3,has:3,have:3,header:3,held:3,henc:3,here:3,herebi:3,herein:3,hereinaft:3,holder:3,home:1,hope:3,how:[1,3],howev:3,idea:3,identifi:3,imag:0,implement:3,impli:3,impos:3,inabl:3,inaccur:3,inc:3,incident:3,includ:3,incompat:3,incorpor:3,independ:3,index:1,indic:3,indirectli:3,induc:3,inform:3,infring:3,inlin:3,insist:3,instal:[1,3],instead:3,intact:3,integr:3,intend:3,intent:3,interchang:3,interest:3,interfac:3,introduc:3,invalid:3,invok:3,irrevers:3,isol:3,issu:[1,3],its:3,itself:3,jame:3,job:3,judgment:3,just:2,keep:3,kernel:3,kind:3,knob:3,know:3,languag:3,larg:3,later:3,latter:3,law:3,lax:3,layout:3,least:[2,3],legal:3,length:3,less:3,lesser:3,lgpl:3,liabl:3,librari:[0,3],licens:1,license:3,licensor:3,limit:3,line:3,link:3,linux:3,list:3,littl:3,loss:3,machin:3,macro:3,made:3,mai:3,mail:3,major:3,make:3,makefil:2,mani:3,materi:3,mean:3,meaning:3,mechan:3,medium:3,meet:3,merchant:3,mere:3,merlin:2,method:3,might:3,modif:3,modifi:3,modul:[1,3],more:3,most:3,must:3,name:3,necessari:3,necessarili:3,need:[2,3],newer:3,non:3,normal:3,note:2,noth:3,notic:3,number:3,numer:3,object:3,oblig:3,obtain:3,ocaml:[0,2],ocamlformat:2,ocamlpro:[0,3],occas:3,odoc:2,offer:3,onc:3,one:3,onli:[2,3],opam:1,oper:3,opt:3,option:3,order:3,ordinari:3,origin:3,other:3,otherwis:3,our:3,out:3,output:3,outsid:3,over:3,own:3,packag:[2,3],page:1,pai:3,paper:3,paramet:3,part:3,parti:3,particular:[2,3],pass:3,patent:3,peopl:3,perform:3,permiss:3,permit:3,pertin:3,physic:3,piec:3,place:3,plu:3,pointer:3,portion:3,pose:3,possibl:3,practic:3,preambl:3,precis:3,prefer:3,prepar:3,present:3,preserv:3,presid:3,price:3,problem:3,produc:3,program:3,programm:3,prohibit:3,promin:3,promot:3,properli:3,properti:3,proprietari:3,protect:3,prove:3,provid:[2,3],publicli:3,publish:3,purpos:3,qualiti:3,quit:3,random:3,rang:3,rare:3,rather:3,readabl:3,reason:3,receiv:3,recipi:3,recommend:3,recompil:3,redistribut:3,refer:3,refrain:3,regardless:3,releas:3,relianc:3,relink:3,remain:3,render:[0,3],repair:3,repositori:2,repres:0,reproduc:3,reput:3,requir:3,respons:3,rest:3,restrict:3,reus:3,revers:3,revis:3,rgb:0,right:3,risk:3,root:3,royalti:3,run:[2,3],safest:3,sai:3,same:3,sampl:3,satisfi:3,school:3,scope:3,script:3,search:1,section:3,see:3,sent:3,separ:3,servic:3,share:3,she:3,should:3,show:3,side:3,sign:3,signatur:3,signific:3,similar:3,simultan:3,sinc:3,singl:3,small:3,softwar:3,sole:3,some:3,someon:3,sometim:3,sourc:[2,3],speak:3,special:3,specifi:3,sphinx:2,spirit:3,squar:3,standard:3,start:3,state:3,statu:3,step:3,still:3,storag:3,straightforwardli:3,strategi:3,street:3,structur:3,subject:3,sublicens:3,subsect:3,subsequ:3,successor:3,suggest:3,suitabl:3,suppli:3,sure:3,surrend:3,sustain:3,system:3,tabl:3,take:3,target:2,ten:3,term:3,termin:3,test:2,text:0,than:3,thei:3,them:3,themselv:3,therefor:3,thi:3,thing:3,think:3,third:3,thoroughli:3,those:3,though:3,threat:3,three:3,threshold:3,through:3,thu:3,time:3,togeth:3,too:3,tool:3,transfer:3,translat:3,tweak:3,two:3,typic:3,uncombin:3,under:3,understood:3,unenforc:3,unicod:0,unless:3,unmodifi:3,unrestrict:3,usa:3,use:3,used:3,useful:3,user:3,uses:3,using:[2,3],util:3,valid:3,variant:3,verbatim:3,veri:3,verifi:3,version:[2,3],vice:3,volum:3,wai:3,want:3,warranti:3,well:3,were:3,what:3,whatev:3,when:3,where:3,wherea:3,wherewith:3,whether:3,which:3,who:3,whole:3,whose:3,why:3,wide:3,widest:3,willing:3,wish:3,without:3,work:3,would:3,write:3,written:3,wrote:3,year:3,you:[2,3],your:[2,3],yoyodyn:3},titles:["About","Welcome to ez_text_renderer","How to install","Copyright and License"],titleterms:{about:0,author:0,build:2,copyright:3,dune:2,ez_text_render:1,how:2,indic:1,instal:2,licens:3,opam:2,tabl:1,welcom:1}})

src/ez_text_renderer/stubs.c
/**************************************************************************/
/*                                                                        */
/*  Copyright (c) 2021 OCamlPro                                           */
/*                                                                        */
/*  All rights reserved.                                                  */
/*  This file is distributed under the terms of the GNU Lesser General    */
/*  Public License version 2.1, with the special exception on linking     */
/*  described in the LICENSE.md file in the root directory.               */
/*                                                                        */
/*                                                                        */
/**************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>

#define CAML_NAME_SPACE
#include "caml/mlvalues.h"
#include "caml/memory.h"
#include "caml/alloc.h"
#include "caml/fail.h"

#include "text_renderer.h"

static void
check_error(
  int result)
{
  switch (result) {
    case TR_OK:
      break;
    case TR_GENERIC_ERROR:
      caml_failwith("Generic error");
      break;
    case TR_OUT_OF_MEMORY:
      caml_raise_out_of_memory();
      break;
    case TR_INITIALIZATION_FAILED:
      caml_failwith("Library initialization failed");
      break;
    case TR_NOT_INITIALIZED:
      caml_failwith("Library not initialized");
      break;
    case TR_UNSUPPORTED_FONT_FORMAT :
      caml_failwith("Unsupported font format");
      break;
    case TR_ERROR_OPENING_FONT:
      caml_failwith("Error opening font");
      break;
    case TR_FONT_NOT_UNICODE:
      caml_failwith("Font does not support Unicode");
      break;
    case TR_UNABLE_TO_SET_SIZE:
      caml_failwith("Unable to set specified font size");
      break;
    case TR_FONT_NOT_SET:
      caml_failwith("The font must be set before calling this function");
      break;
    case TR_CANT_OPEN_FILE:
      caml_failwith("Unable to open file");
      break;
    default:
      caml_failwith("Unknown error");
      break;
  }
}

static bool
int_in_range(
  int i,
  int min,
  int max)
{
  if (max < min) {
    int t = max;
    max = min;
    min = t;
  }
  return ((i >= min) && (i <= max));
}

static uint32_t
Color_val(
  value mlColor)
{
  CAMLparam1(mlColor);
  int b = Int_val(Field(mlColor, 0));
  int g = Int_val(Field(mlColor, 1));
  int r = Int_val(Field(mlColor, 2));
  int a = Int_val(Field(mlColor, 3));
  if (!int_in_range(a, 0, 0xFF))
    caml_invalid_argument("Alpha component not in range 0-255");
  if (!int_in_range(r, 0, 0xFF))
    caml_invalid_argument("Red component not in range 0-255");
  if (!int_in_range(g, 0, 0xFF))
    caml_invalid_argument("Green component not in range 0-255");
  if (!int_in_range(b, 0, 0xFF))
    caml_invalid_argument("Blue component not in range 0-255");
  uint32_t color = argb_to_int(a, r, g, b);
  CAMLreturnT(uint32_t, color);
}

CAMLprim value
ml_tr_init(
  value mlUnit)
{
  CAMLparam1(mlUnit);
  int result = tr_init();
  check_error(result);
  CAMLreturn(Val_unit);
}

CAMLprim value
ml_tr_release(
  value mlUnit)
{
  CAMLparam1(mlUnit);
  int result = tr_release();
  check_error(result);
  CAMLreturn(Val_unit);
}

CAMLprim value
ml_tr_set_font(
  value mlFontFile,
  value mlFontHeight)
{
  CAMLparam2(mlFontFile, mlFontHeight);
  const char *font_file = String_val(mlFontFile);
  int height = Int_val(mlFontHeight);
  int result = tr_set_font(font_file, height);
  check_error(result);
  CAMLreturn(Val_unit);
}

CAMLprim value
ml_tr_compute_text_width(
  value mlText)
{
  CAMLparam1(mlText);
  const char *text = String_val(mlText);
  int width = 0;
  int result = tr_compute_text_width(text, &width);
  check_error(result);
  CAMLreturn(Val_int(width));
}

CAMLprim value
ml_tr_render_text(
  value mlText,
  value mlFrontColor,
  value mlBackColor,
  value mlOutArea,
  value mlOutArray,
  value mlOutDimensions)
{
  CAMLparam5(mlText, mlFrontColor, mlBackColor, mlOutArea, mlOutArray);
  CAMLxparam1(mlOutDimensions);
  CAMLlocal1(mlColor);

  const char *text = String_val(mlText);
  uint32_t front_color = Color_val(mlFrontColor);
  uint32_t back_color = Color_val(mlBackColor);
  int x = Int_val(Field(mlOutArea, 0));
  int y = Int_val(Field(mlOutArea, 1));
  int width = Int_val(Field(mlOutArea, 2));
  int height = Int_val(Field(mlOutArea, 3));
  int dest_width = Int_val(Field(mlOutDimensions, 0));
  int dest_height = Int_val(Field(mlOutDimensions, 1));

  int data_size = width * height * BYTES_PER_PIXEL;
  uint8_t *data = malloc(data_size * sizeof(uint8_t));
  if (data == NULL) {
    caml_raise_out_of_memory();
  }

  int result =
    tr_render_text(text, front_color, back_color,
                   0, 0, width, height, data, width, height);
  if (result != TR_OK) {
    free(data);
    check_error(result);
  }

  for (int i = 0; (i < height) && (y + i < dest_height); ++i) {
    for (int j = 0; (j < width) && (x + j < dest_width); ++j) {
      int src_idx = (i * width + j) * BYTES_PER_PIXEL;
      int dst_idx = (y + i) * dest_width + (x + j);
      mlColor = caml_alloc_tuple(4);
      Store_field(mlColor, 0, Val_int(data[src_idx + 0])); // blue
      Store_field(mlColor, 1, Val_int(data[src_idx + 1])); // green
      Store_field(mlColor, 2, Val_int(data[src_idx + 2])); // red
      Store_field(mlColor, 3, Val_int(data[src_idx + 3])); // alpha
      Store_field(mlOutArray, dst_idx, mlColor);
    }
  }

  free(data);
  CAMLreturn(Val_unit);
}

CAMLprim value
ml_tr_render_text_bc(value *a, int n)
{
  return ml_tr_render_text(a[0],a[1],a[2],a[3],a[4],a[5]);
}

CAMLprim value
ml_tr_dump_image(
  value mlArray,
  value mlDimensions,
  value mlOutFile)
{
  CAMLparam3(mlArray, mlDimensions, mlOutFile);
  CAMLlocal1(mlColor);

  int width = Int_val(Field(mlDimensions, 0));
  int height = Int_val(Field(mlDimensions, 1));
  const char *outfile = String_val(mlOutFile);

  int data_size = width * height * BYTES_PER_PIXEL;
  uint8_t *data = malloc(data_size * sizeof(uint8_t));
  if (data == NULL) {
    caml_raise_out_of_memory();
  }

  for (int i = 0; i < height; ++i) {
    for (int j = 0; j < width; ++j) {
      int src_idx = i * width + j;
      int dst_idx = src_idx * BYTES_PER_PIXEL;
      mlColor = Field(mlArray, src_idx);
      data[dst_idx + 0] = Int_val(Field(mlColor, 0)); // blue
      data[dst_idx + 1] = Int_val(Field(mlColor, 1)); // green
      data[dst_idx + 2] = Int_val(Field(mlColor, 2)); // red
      data[dst_idx + 3] = Int_val(Field(mlColor, 3)); // alpha
    }
  }

  int result = tr_dump_image(data, width, height, outfile);
  if (result != TR_OK) {
    free(data);
    check_error(result);
  }

  free(data);
  CAMLreturn(Val_unit);
}


src/ez_text_renderer/text_renderer.h
/**************************************************************************/
/*                                                                        */
/*  Copyright (c) 2021 OCamlPro                                           */
/*                                                                        */
/*  All rights reserved.                                                  */
/*  This file is distributed under the terms of the GNU Lesser General    */
/*  Public License version 2.1, with the special exception on linking     */
/*  described in the LICENSE.md file in the root directory.               */
/*                                                                        */
/*                                                                        */
/**************************************************************************/

#include <stdint.h>

#define BYTES_PER_PIXEL 4

#define alpha_of_int(i) (((i) & 0xFF000000) >> 24)
#define red_of_int(i)   (((i) & 0x00FF0000) >> 16)
#define green_of_int(i) (((i) & 0x0000FF00) >>  8)
#define blue_of_int(i)  (((i) & 0x000000FF) >>  0)

#define rgb_to_int(r,g,b) \
  (((      0xFF) << 24) | \
   (((r) & 0xFF) << 16) | \
   (((g) & 0xFF) <<  8) | \
   (((b) & 0xFF) <<  0))

#define argb_to_int(a,r,g,b) \
  ((((a) & 0xFF) << 24) | \
   (((r) & 0xFF) << 16) | \
   (((g) & 0xFF) <<  8) | \
   (((b) & 0xFF) <<  0))

#define TR_OK                       0
#define TR_GENERIC_ERROR            1
#define TR_OUT_OF_MEMORY            2
#define TR_INITIALIZATION_FAILED    3
#define TR_NOT_INITIALIZED          4
#define TR_UNSUPPORTED_FONT_FORMAT  5
#define TR_ERROR_OPENING_FONT       6
#define TR_FONT_NOT_UNICODE         7
#define TR_UNABLE_TO_SET_SIZE       8
#define TR_FONT_NOT_SET             9
#define TR_CANT_OPEN_FILE           10

int tr_init();
int tr_release();
int tr_set_font(const char *font_file, int height);
int tr_compute_text_width(const char *mb_text, int *out_width);
int tr_render_text(const char *mb_text, uint32_t front_color, uint32_t back_color, int x, int y, int w, int h, uint8_t *data, int dw, int dh);
int tr_dump_image(const uint8_t *data, int dw, int dh, const char *outfile);


src/ez_text_renderer/ezTextRenderer.ml
(**************************************************************************)
(*                                                                        *)
(*  Copyright (c) 2021 OCamlPro                                           *)
(*                                                                        *)
(*  All rights reserved.                                                  *)
(*  This file is distributed under the terms of the GNU Lesser General    *)
(*  Public License version 2.1, with the special exception on linking     *)
(*  described in the LICENSE.md file in the root directory.               *)
(*                                                                        *)
(*                                                                        *)
(**************************************************************************)

type position = {
  x : int;
  y : int;
}

type dimension = {
  width : int;
  height : int;
}

type area = {
  x : int;
  y : int;
  width : int;
  height : int;
}

type color = {
  blue : int;
  green : int;
  red : int;
  alpha : int;
}

external init : unit -> unit = "ml_tr_init"

external release : unit -> unit = "ml_tr_release"

external set_font : string -> int -> unit = "ml_tr_set_font"

external compute_text_width : string -> int = "ml_tr_compute_text_width"

external render_text : string -> color -> color -> area -> color array -> dimension -> unit = "ml_tr_render_text_bc" "ml_tr_render_text"

external dump_image : color array -> dimension -> string -> unit = "ml_tr_dump_image"


src/ez_text_renderer/text_renderer.c
/**************************************************************************/
/*                                                                        */
/*  Copyright (c) 2021 OCamlPro                                           */
/*                                                                        */
/*  All rights reserved.                                                  */
/*  This file is distributed under the terms of the GNU Lesser General    */
/*  Public License version 2.1, with the special exception on linking     */
/*  described in the LICENSE.md file in the root directory.               */
/*                                                                        */
/*                                                                        */
/**************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <wchar.h>
#include <locale.h>
#include <ft2build.h>
#include FT_FREETYPE_H

#include "text_renderer.h"

#define min(a,b) ((a) <= (b) ? (a) : (b))
#define max(a,b) ((a) >= (b) ? (a) : (b))

#define alpha_blend(alpha,src,dst) \
  (((alpha) * (src) + (0xFF - (alpha)) * (dst)) / 0xFF)

static bool initialized = false;
static bool font_set = false;
static FT_Library library;
static FT_Face face;

static wchar_t *
mb_to_wc(
  const char *mb_text)
{
  size_t len = mbstowcs(NULL, mb_text, 0) + 1;
  wchar_t *wc_text = malloc(len * sizeof(wchar_t));
  if (wc_text != NULL) {
    mbstowcs(wc_text, mb_text, len);
  }
  return wc_text;
}

int
tr_init()
{
  int error;

  if (initialized) {
    return TR_OK;
  }

  setlocale(LC_ALL, "");

  error = FT_Init_FreeType(&library);
  if (error) {
    return TR_INITIALIZATION_FAILED;
  }

  initialized = true;

  return TR_OK;
}

int
tr_release()
{
  if (!initialized) {
    return TR_OK;
  }

  FT_Done_FreeType(library);

  initialized = false;

  return TR_OK;
}

int
tr_set_font(
  const char *font_file,
  int height)
{
  int error;

  if (!initialized) {
    return TR_NOT_INITIALIZED;
  }

  error = FT_New_Face(library, font_file, 0 /* face_index */, &face);
  if (error == FT_Err_Unknown_File_Format) {
    return TR_UNSUPPORTED_FONT_FORMAT;
  }
  else if (error) {
    return TR_ERROR_OPENING_FONT;
  }

  error = FT_Select_Charmap(face, FT_ENCODING_UNICODE);
  if (error) {
    return TR_FONT_NOT_UNICODE;
  }

  int try_height = height;
  do {
    FT_Set_Pixel_Sizes(face, 0, try_height--);
    if (error) {
      return TR_UNABLE_TO_SET_SIZE;
    }
  } while ((height <
	    (face->size->metrics.ascender +
	     labs(face->size->metrics.descender)) >> 6) ||
           (height <
	    (face->size->metrics.height) >> 6));

  font_set = true;

/*
  FT_Size_RequestRec req = { FT_SIZE_REQUEST_TYPE_BBOX, 0, height * 64,0, 0 };
  FT_Request_Size(face, &req);
*/
/*
  printf("num_faces: %ld\n", face->num_faces);
  printf("face_index: %ld\n", face->face_index);
  printf("face_flags: %ld\n", face->face_flags);
  printf("num_glyphs: %ld\n", face->num_glyphs);
  printf("family_name: %s\n", face->family_name);
  printf("style_name: %s\n", face->style_name);
  printf("num_fixed_sizes: %d\n", face->num_fixed_sizes);
  printf("num_charmaps: %d\n", face->num_charmaps);

  printf("bbox.xMin: %ld\n", face->bbox.xMin >> 6);
  printf("bbox.yMin: %ld\n", face->bbox.yMin >> 6);
  printf("bbox.xMax: %ld\n", face->bbox.xMax >> 6);
  printf("bbox.yMax: %ld\n", face->bbox.yMax >> 6);
  printf("units_per_EM: %d\n", face->units_per_EM >> 6);
  printf("ascender: %d\n", face->ascender >> 6);
  printf("descender: %d\n", face->descender >> 6);
  printf("height: %d\n", face->height >> 6);

  printf("metrics.ascender : %ld\n", face->size->metrics.ascender >> 6);
  printf("metrics.descender : %ld\n", face->size->metrics.descender >> 6);
  printf("metrics.height : %ld\n", face->size->metrics.height >> 6);
*/
  return TR_OK;
}


int
tr_compute_text_width(
  const char *mb_text,
  int *out_width)
{
  if (!initialized) {
    return TR_NOT_INITIALIZED;
  }

  if (!font_set) {
    return TR_FONT_NOT_SET;
  }

  wchar_t *wc_text = mb_to_wc(mb_text);
  if (wc_text == NULL) {
    return TR_OUT_OF_MEMORY;
  }

  size_t len = wcslen(wc_text);

  int width = 0;
  for (size_t i = 0; i < len; ++i) {

    int error = FT_Load_Char(face, wc_text[i], FT_LOAD_DEFAULT);
    if (error) {
      continue; // we just ignore character errors
    }

    width += face->glyph->advance.x >> 6; // should adjust last character
  }

  *out_width = width;

  free(wc_text);

  return TR_OK;
}


static void
tr_draw_glyph(
  FT_GlyphSlot glyph,
  uint32_t color,
  int pen_x,
  int pen_y,
  int x,
  int y,
  int w,
  int h,
  uint8_t *dest,
  int dw,
  int dh)
{
  int lim_x = min(x + w, dw);
  int lim_y = min(y + h, dh);

  uint8_t a = alpha_of_int(color);
  uint8_t r = red_of_int(color);
  uint8_t g = green_of_int(color);
  uint8_t b = blue_of_int(color);

  for (int i = 0; i < glyph->bitmap.rows; ++i) {

    int di = pen_y - glyph->bitmap_top + i;
    if (di < 0) continue;
    else if (di >= lim_y) break;

    for (int j = 0; j < glyph->bitmap.width; ++j) {

      int dj = pen_x + glyph->bitmap_left + j;
      if (dj < 0) continue;
      else if (dj >= lim_x) break;

      int s_idx = i * glyph->bitmap.pitch + j;
      uint8_t alpha = glyph->bitmap.buffer[s_idx] * a / 0xFF;

      int d_idx;

      switch (alpha) {

        case 0x00:
          // Fully transparent pixel: don't bother drawing
          break;

        case 0xFF:
          // Fully opaque pixel: no alpha computation
          d_idx = (di * dw + dj) * BYTES_PER_PIXEL;
          dest[d_idx + 0] = b;
          dest[d_idx + 1] = g;
          dest[d_idx + 2] = r;
          dest[d_idx + 3] = 0xFF;
          break;

        default:
          // Semi-transparent pixel: apply alpha
          d_idx = (di * dw + dj) * BYTES_PER_PIXEL;
          dest[d_idx + 0] = alpha_blend(alpha, b, dest[d_idx + 0]);
          dest[d_idx + 1] = alpha_blend(alpha, g, dest[d_idx + 1]);
          dest[d_idx + 2] = alpha_blend(alpha, r, dest[d_idx + 2]);
          dest[d_idx + 3] = 0xFF;
      }
    }
  }
}

int tr_render_text(
  const char *mb_text,
  uint32_t front_color,
  uint32_t back_color,
  int x,
  int y,
  int w,
  int h,
  uint8_t *data,
  int dw,
  int dh)
{
  if (!initialized) {
    return TR_NOT_INITIALIZED;
  }

  if (!font_set) {
    return TR_FONT_NOT_SET;
  }

  wchar_t *wc_text = mb_to_wc(mb_text);
  if (wc_text == NULL) {
    return TR_OUT_OF_MEMORY;
  }

  int r = red_of_int(back_color);
  int g = green_of_int(back_color);
  int b = blue_of_int(back_color);

  for (int i = 0; (i < h) && (y + i < dh); ++i) {
    for (int j = 0; (j < w) && (x + j < dw); ++j) {
      int d_idx = ((y + i) * dw + (x + j)) * BYTES_PER_PIXEL;
      data[d_idx + 0] = b;
      data[d_idx + 1] = g;
      data[d_idx + 2] = r;
      data[d_idx + 3] = 0xFF; // alpha ignored on background color
    }
  }

  int pen_x = x;
  int pen_y = y + (face->size->metrics.ascender >> 6);

  size_t len = wcslen(wc_text);

  for (size_t i = 0; i < len; ++i) {

    int error = FT_Load_Char(face, wc_text[i], FT_LOAD_RENDER);
    if (error) {
      continue; // we just ignore character errors
    }

    // FT_PIXEL_MODE_MONO = might be used
    // FT_PIXEL_MODE_BGRA = rare, used by Google
    // FT_PIXEL_MODE_GRAY2/GRAY4 = even rarer
    // FT_PIXEL_MODE_LCD/LCD_V = ???
    if (face->glyph->bitmap.pixel_mode != FT_PIXEL_MODE_GRAY) {
      fprintf(stderr, "Unsupported pixel mode %d\n",
              face->glyph->bitmap.pixel_mode);
      continue; // should try to support others
    }

    tr_draw_glyph(face->glyph, front_color,
                  pen_x, pen_y, x, y, w, h, data, dw, dh);

    pen_x += face->glyph->advance.x >> 6;
  }

  free(wc_text);

  return TR_OK;
}

int
tr_dump_image(
  const uint8_t *data,
  int dw,
  int dh,
  const char *outfile)
{
  FILE *out = fopen(outfile, "wb");
  if (out == NULL) {
    return TR_CANT_OPEN_FILE;
  }

  fprintf(out, "P6\n%d %d\n255\n", dw, dh);

  for (int i = 0; i < dw * dh * BYTES_PER_PIXEL; i += BYTES_PER_PIXEL) {
    fputc(data[i + 2], out); // red
    fputc(data[i + 1], out); // green
    fputc(data[i + 0], out); // blue
  }

  fclose(out);

  return TR_OK;
}


